File: sketch/src/main/java/me/xiaopan/sketch/state/DrawableStateImage.java
/*
 * Copyright (C) 2016 Peng fei Pan <sky@xiaopan.me>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.xiaopan.sketch.state;

import android.content.Context;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;

import me.xiaopan.sketch.drawable.ShapeBitmapDrawable;
import me.xiaopan.sketch.request.DisplayOptions;
import me.xiaopan.sketch.request.ImageViewInterface;
import me.xiaopan.sketch.request.ShapeSize;
import me.xiaopan.sketch.shaper.ImageShaper;

/**
 * 给什么图片显示什么图片，支持ShapeSize和ImageShaper
 */
@SuppressWarnings("unused")
public class DrawableStateImage implements StateImage {
    private Drawable originDrawable;
    private int resId = -1;

    public DrawableStateImage(Drawable drawable) {
        this.originDrawable = drawable;
    }

    public DrawableStateImage(int resId) {
        this.resId = resId;
    }

    @Override
    public Drawable getDrawable(Context context, ImageViewInterface imageViewInterface, DisplayOptions displayOptions) {
        Drawable drawable = originDrawable;
        if (drawable == null && resId != -1) {
            drawable = context.getResources().getDrawable(resId);
        }

        ShapeSize shapeSize = displayOptions.getShapeSize();
        ImageShaper imageShaper = displayOptions.getImageShaper();
        if ((shapeSize != null || imageShaper != null) && drawable != null
                && drawable instanceof BitmapDrawable) {
            drawable = new ShapeBitmapDrawable(context, (BitmapDrawable) drawable, shapeSize, imageShaper);
        }

        return drawable;
    }

    public Drawable getOriginDrawable() {
        return originDrawable;
    }

    public int getResId() {
        return resId;
    }
}

File: sketch-core/src/androidMain/kotlin/com/github/panpf/sketch/state/DrawableStateImage.kt
/*
 * Copyright (C) 2024 panpf <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.panpf.sketch.state

import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import androidx.annotation.DrawableRes
import com.github.panpf.sketch.Image
import com.github.panpf.sketch.Sketch
import com.github.panpf.sketch.asImage
import com.github.panpf.sketch.drawable.DrawableFetcher
import com.github.panpf.sketch.drawable.EquitableDrawable
import com.github.panpf.sketch.drawable.RealDrawableFetcher
import com.github.panpf.sketch.drawable.ResDrawableFetcher
import com.github.panpf.sketch.drawable.asEquitable
import com.github.panpf.sketch.request.ImageRequest

/**
 * [EquitableDrawable] as [StateImage]
 *
 * @see com.github.panpf.sketch.core.android.test.state.DrawableStateImageTest.testAsStateImage
 */
fun EquitableDrawable.asStateImage(): DrawableStateImage = DrawableStateImage(this)

/**
 * [Drawable] as [StateImage]
 *
 * @see com.github.panpf.sketch.core.android.test.state.DrawableStateImageTest.testAsStateImage
 */
fun Drawable.asStateImage(equalityKey: Any): DrawableStateImage =
    DrawableStateImage(RealDrawableFetcher(this.asEquitable(equalityKey)))

/**
 * [ColorDrawable] as [StateImage]
 *
 * @see com.github.panpf.sketch.core.android.test.state.DrawableStateImageTest.testAsStateImage
 */
fun ColorDrawable.asStateImage(): DrawableStateImage =
    DrawableStateImage(RealDrawableFetcher(this.asEquitable()))

/**
 * Create a [DrawableStateImage] with [EquitableDrawable]
 *
 * @see com.github.panpf.sketch.core.android.test.state.DrawableStateImageTest.testDrawableStateImage
 */
fun DrawableStateImage(drawable: EquitableDrawable): DrawableStateImage =
    DrawableStateImage(RealDrawableFetcher(drawable))

/**
 * Create a [DrawableStateImage] with resource drawable
 *
 * @see com.github.panpf.sketch.core.android.test.state.DrawableStateImageTest.testDrawableStateImage
 */
fun DrawableStateImage(@DrawableRes resId: Int): DrawableStateImage =
    DrawableStateImage(ResDrawableFetcher(resId))

/**
 * Use [Drawable] as the state [Drawable]
 *
 * @see com.github.panpf.sketch.core.android.test.state.DrawableStateImageTest.testDrawableStateImage
 */
data class DrawableStateImage(
    val drawableFetcher: DrawableFetcher
) : StateImage {

    override val key: String = "Drawable(${drawableFetcher.key})"

    override fun getImage(
        sketch: Sketch,
        request: ImageRequest,
        throwable: Throwable?
    ): Image? {
        return try {
            drawableFetcher.getDrawable(request.context)
        } catch (e: Throwable) {
            sketch.logger.w("DrawableStateImage. getDrawable error. ${e.message}")
            e.printStackTrace()
            null
        }?.asImage()
    }

    override fun toString(): String = "DrawableStateImage(drawable=$drawableFetcher)"
}